src/zenml/integrations/kubernetes/orchestrators/dag_runner.py

```diff
@@ -0,0 +1,361 @@
+            Active nodes.
+        """
+        return [
+            node
+            for node in self.nodes.values()
+            if node.status in {NodeStatus.RUNNING, NodeStatus.STARTING}
+        ]
+
+    def _initialize_startup_queue(self) -> None:
+        """Initialize the startup queue.
+
+        The startup queue contains all nodes that are ready to be started.
+        """
+        for node in self.nodes.values():
+            if node.status in {NodeStatus.READY, NodeStatus.STARTING}:
+                self.startup_queue.put(node)
+
+    def _can_start_node(self, node: Node) -> bool:
+        """Check if a node can be started.
+
+        Args:
+            node: The node to check.
+
+        Returns:
+            Whether the node can be started.
+        """
+        return all(
+            self.nodes[upstream_node_id].status == NodeStatus.COMPLETED
+            for upstream_node_id in node.upstream_nodes
+        )
+
+    def _should_skip_node(self, node: Node) -> bool:
+        """Check if a node should be skipped.
+
+        Args:
+            node: The node to check.
+
+        Returns:
+            Whether the node should be skipped.
+        """
+        return any(
+            self.nodes[upstream_node_id].status
+            in {NodeStatus.FAILED, NodeStatus.SKIPPED, NodeStatus.CANCELLED}
+            for upstream_node_id in node.upstream_nodes
+        )
+
+    def _start_node(self, node: Node) -> None:
+        """Start a node.
+
+        This will start of a thread that will run the startup function.
+
+        Args:
+            node: The node to start.
+        """
+        node.status = NodeStatus.STARTING
+
+        def _start_node_task() -> None:
+            if self.shutdown_event.is_set():
+                logger.debug(
+                    "Cancelling startup of node `%s` because shutdown was "
+                    "requested.",
+                    node.id,
+                )
+                return
+
+            try:
+                node.status = self.node_startup_function(node)
+            except Exception:
+                node.status = NodeStatus.FAILED
+                logger.exception("Node `%s` failed to start.", node.id)
+            else:
+                logger.info(
+                    "Node `%s` started (status: %s)", node.id, node.status
+                )
+
+        self.startup_executor.submit(_start_node_task)
+
+    def _stop_node(self, node: Node) -> None:
+        """Stop a node.
+
+        Args:
+            node: The node to stop.
+
+        Raises:
+            RuntimeError: If the node stop function is not set.
+        """
+        if not self.node_stop_function:
+            raise RuntimeError("Node stop function is not set.")
+
+        self.node_stop_function(node)
+
+    def _stop_all_nodes(self) -> None:
+        """Stop all running nodes."""
+        for node in self.running_nodes:
+            self._stop_node(node)
+            node.status = NodeStatus.CANCELLED
```

Comment:
This can potentially create a flood of external calls to `node_stop_function` if you have a large number of steps running at the same time. You should try to protect this by adding some configurable rate-limiting measures here (max requests per second, or interval).